### PR TITLE
[DI] Add logging and better failure recovery to AutowirePass

### PR DESCRIPTION
**Pull request by nicolas-grekas**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | -
| License       | MIT
| Doc PR        | -

So useful to understand what autowiring is doing.

### CODE REVIEW

**Comment by fabpot** [comment]
👍

File: src/Symfony/Component/DependencyInjection/Compiler/AutowirePass.php [hunk 1]
```diff
@@ -449,24 +454,53 @@ private function createAutowiredDefinition(\ReflectionClass $typeHint)
...
 
         $this->populateAvailableType($argumentId, $argumentDefinition);
 
-        $this->processValue($argumentDefinition, true);
-        $this->currentId = $currentId;
+        try {
+            $this->processValue($argumentDefinition, true);
+            $this->container->setDefinition($argumentId, $argumentDefinition);
+        } catch (RuntimeException $e) {
+            // revert any changes done to our internal state
```
Comment (stof) [inline src/Symfony/Component/DependencyInjection/Compiler/AutowirePass.php]: can we have tests covering this ?

Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Compiler/AutowirePass.php]: test case added

File: src/Symfony/Component/DependencyInjection/Compiler/AutowirePass.php [hunk 2]
```diff
@@ -449,24 +454,53 @@ private function createAutowiredDefinition(\ReflectionClass $typeHint)
...
         }
 
+        $ambiguousServiceTypes = $this->ambiguousServiceTypes;
+        $currentDefinition = $this->currentDefinition;
+        $definitions = $this->container->getDefinitions();
         $currentId = $this->currentId;
         $this->currentId = $argumentId = sprintf('autowired.%s', $typeHint->name);
-
-        $argumentDefinition = $this->container->register($argumentId, $typeHint->name);
+        $this->currentDefinition = $argumentDefinition = (new Definition($typeHint->name));
```
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Compiler/AutowirePass.php]: removed

**Comment by fabpot** [comment]
Thank you @nicolas-grekas.

**Pull request merged**